### ISSUE TITLE
pytorch study 5주차

### ISSUE THREAD

**Issue by hyunsu-yang** [hyunsu-yang/PyTorch#5]
## 강의
- https://www.edwith.org/boostcourse-dl-pytorch/joinLectures/24016

## 목차 
- Lab-09-2 Weight initialization
- Lab-09-3 Dropout
- Lab-09-4 Batch Normalization

**Comment by hyunsu-yang** [comment]
## Lab-09-2 Weight initialization
### 핵심 키워드&요약
- Weight initialization  

### 문제
#### Weight initialization 이 중요한 이유와 초기화 방법에는 어떤 것이 있는지 기재해주세요.
- W 초기화 방식에 따라 성능에 큰 영향
- 가중치 초기화만 적절히 해줘도 기울기 소실 문제과 같은 문제를 완화시킬수 있음.
- 0 초기화 는 매우 안좋음.  backpropagation 시 gradient 계산이 안되어 학습이 안됨
- 초기화 방법 두가지
  - 세이비어 초기화(Xavier Initialization)
    - 이전 층의 뉴런 개수와 다음 층의 뉴런 개수를 모두 반영
    - activation function으로 시그모이드 함수나 하이퍼볼릭탄젠트 함수를 같이 사용.
  - He 초기화(He initialization)
    - 다음 층의 뉴런의 수를 반영하지 않음.
    - activation function으로 ReLU 함수를 같이 사용.

## Lab-09-3 Dropout
### 핵심 키워드&요약
- 과적적화(Overfitting)
- 드롭아웃(Dropout) 

### 문제
#### Overfitting을 방지하는 Dropout에 대한 설명과 주의사항에 대해 기재해주세요. 
-  학습 과정에서 신경망의 일부를 사용하지 않는 방법
- 학습 시에 인공 신경망이 특정 뉴런 또는 특정 조합에 너무 의존적이게 되는 것을 방지해주고,
 매번 랜덤 선택으로 뉴런들을 사용하지 않으므로 서로 다른 신경망들을 앙상블하여 사용하는 것 
같은 효과를 내어 과적합을 방지
- 드롭아웃은 신경망 학습 시에만 사용하고, 예측 시에는 사용하지 않는 것이 일반
![image](https://user-images.githubusercontent.com/19383191/94404924-b81e6580-01aa-11eb-9999-34d354543ad0.png)

#### Overfitting을 방지하기 위해 Dropout 이외에도 다른 방법이 있다면 생각나는데로 기재해주세요
- 데이터 양 늘리기
  - 모델은 데이터의 양이 적을 경우, 해당 데이터의 특정 패턴이나 노이즈까지 쉽게 암기하기 되므로
    과적합 현상이 발생할 확률이 늘어남.
- 모델의 복잡도 줄이기
  - 은닉층(hidden layer)의 수나 매개변수의 수 를 줄이기


## Lab-09-4 Batch Normalization
### 핵심 키워드&요약
- internal Covariate shift
  - 한 레이어 마다 입력과 출력 분포값이 달라지는 
  - 레이어가 많아질 때마다 더 변화가 심해지는
- Batch Normalization
  - 배치 정규화는 인공 신경망의 각 층에 들어가는 입력을 평균과 분산으로 정규화하여 학습을 효율적으로 만듬.
- 경사 소실(Gradient Vanishing) 
  - 간접 해결
    - change activation function => Relu
    - careful initialization
    - small learning rate
  - 직접 해결
    - Batch Normalization
      - internal Covariate shift 문제를 해결하기 위해 각 레이어 마다 노멀라이제이션 해주는거..
- 폭발(Exploding)
  - 기울기가 점차 커지더니 가중치들이 비정상적으로 큰 값이 되면서 결국 발산
  - Gradient Vanishing 반대
### 문제
#### Internal Covariate Shift 가 무엇인지 기재해주세요.
- 학습 과정에서 층 별로 입력 데이터 분포가 달라지는 현상
- 이전 층들의 학습에 의해 이전 층의 가중치 값이 바뀌게 되면, 
  현재 층에 전달되는 입력 데이터의 분포가 현재 층이 학습했던 시점의 분포와 차이가 발생

#### 배치 정규화(Batch Normalization) 필요한 이유와  간략히 수식을 그리고 설명해주세요.
- Gradient Vanishing 와 Exploding 완화하기 위해
![image](https://user-images.githubusercontent.com/19383191/94407608-9de68680-01ae-11eb-8d7c-139e0fcf869d.png)
- 입력에 대해 평균을 0으로 만들고, 정규화를 합니다
- 정규화 된 데이터에 대해서 스케일과 시프트를 수행
-  γ는 스케일을 위해 사용하고, β는 시프트를 하는 것에 사용하며 다음 레이어에 일정한 범위의 값들만 전달

#### 배치 정규화(Batch Normalization)의 장단점을 아는대로 기재해주세요
- 기울기 소실 문제가 크게 개선
- 가중치 초기화에 훨씬 덜 민감해짐.
- 훨씬 큰 학습률을 사용할 수 있어 학습 속도를 개선
- 배치 정규화는 모델을 복잡하게 하며, 추가 계산을 하는 것이므로 
  테스트 데이터에 대한 예측 시에 실행 시간이 느려짐. 그래서 서비스 속도를 고려하는 관점에서는
   배치 정규화가 꼭 필요한지 고민해야 함.

#### 배치 정규화(Batch Normalization)에는 몇가지 한계가 있다고 합니다. 아는대로 기재해주세요.
- 미니 배치 크기에 의존적이다.
  - 너무 작은 배치 크기에서는 잘 동작하지 않음.
- RNN에 적용하기 어려움
  - RNN은 각 시점(time step)마다 다른 통계치를 가지고 있기 때문에..
  - 그래서 layer normalization 등장.
